### PR TITLE
[SPARK-25822][PySpark]Fix a race condition when releasing a Python worker

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

There is a race condition when releasing a Python worker. If `ReaderIterator.handleEndOfDataSection` is not running in the task thread, when a task is early terminated (such as `take(N)`), the task completion listener may close the worker but "handleEndOfDataSection" can still put the worker into the worker pool to reuse.

https://github.com/zsxwing/spark/commit/0e07b483d2e7c68f3b5c3c118d0bf58c501041b7 is a patch to reproduce this issue.

I also found a user reported this in the mail list: http://mail-archives.apache.org/mod_mbox/spark-user/201610.mbox/%3CCAAUq=H+YLUEpd23nwvq13Ms5hOStkhX3ao4f4zQV6sgO5zM-xA@mail.gmail.com%3E

This PR fixes the issue by using `compareAndSet` to make sure we will never return a closed worker to the work pool.

## How was this patch tested?

Jenkins.

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @ueshin

File: core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala [hunk 1]
```diff
@@ -114,7 +114,7 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
 
     context.addTaskCompletionListener[Unit] { _ =>
       writerThread.shutdownOnTaskCompletion()
-      if (!reuseWorker || !released.get) {
+      if (!reuseWorker || released.compareAndSet(false, true)) {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala]: I think `released` variable here indices if `releasePythonWorker` is actually called or not. Do you mind if I ask to elaborate why this `released` it set to `true` here as well?

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala]: I think `released` variable here indices if `releasePythonWorker` is actually called or not before. Shall we rename and update the comments at  `val released`?

Comment (zsxwing) [reply core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala]: Addressed

**Comment by zsxwing** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by ueshin** [comment]
LGTM.

**Comment by ueshin** [comment]
Thanks! merging to master/2.4/2.3.

**Pull request closed without merging**